tools/cldr-code/src/main/java/org/unicode/cldr/util/GlossonymConstructor.java

```diff
@@ -0,0 +1,124 @@
+package org.unicode.cldr.util;
+
+import com.ibm.icu.util.Output;
+import org.unicode.cldr.test.CheckNames;
+
+/**
+ * Automatically construct language names (glossonyms)
+ *
+ * Example: in German (de), for the path
+ *
+ *       //ldml/localeDisplayNames/languages/language[@type="ro_MD"]
+ *
+ * the value "Rumänisch (Republik Moldau)" is automatically constructed based on the code "ro_MD".
+ *
+ * The constructed value is a default if no preferable value is submitted or inherited.
+ * A different (non-constructed) value, such as "Moldauisch", may become the winning
+ * value instead of the constructed value.
+ */
+public class GlossonymConstructor {
+
+    /**
+     * Some paths with this prefix can get automatically constructed values
+     */
+    public static final String PATH_PREFIX = "//ldml/localeDisplayNames/languages/language[@type=\"";
+
+    /**
+     * The code such as "ro_MD" must contain an underscore, otherwise there is no constructed value.
+     * Underscore also serves as a clue for recognizing a value that is from code-fallback; for example,
+     * when the value "ro_MD" is inherited, the underscore implies it's a raw code ("bogus value") and should
+     * be replaced by a constructed value like "Rumänisch (Republik Moldau)"; but when the value "Moldauisch"
+     * (without underscore) is inherited, it should not be replaced by a constructed value
+     */
+    private static final String CODE_SEPARATOR = "_";
+
+    /**
+     * For "pathWhereFound" when the value is constructed.
+     * It is non-null to satisfy TestPathHeadersAndValues.
+     * It should not be treated as an actual path; for example, the
+     * Survey Tool Info Panel should not show a broken "Jump to original" link.
+     */
+    public static final String PSEUDO_PATH = "constructed";
+
+    /**
+     * Is the given path eligible for getting a constructed value?
+     *
+     * @param xpath the given path
+     * @return true if eligible
+     */
+    public static boolean pathIsEligible(String xpath) {
+        return xpath.startsWith(PATH_PREFIX) && xpath.contains(CODE_SEPARATOR);
+    }
+
+    /**
+     * Is the given value bogus, and therefore eligible for getting replaced by a constructed value?
+     *
+     * @param value the given value
+     * @return true if bogus
+     */
+    public static boolean valueIsBogus(String value) {
+        return value == null || value.contains(CODE_SEPARATOR) || CheckNames.matchDigitPattern(value);
```

Comment:
odd layering to have a util call a 'test' - matchDigitPattern probably belongs under util ideally